examples/tox21/tox21_tensorgraph_graph_conv.py

```diff
@@ -18,87 +22,94 @@
-
-  degree_slice = Input(shape=(None, 2), dtype=tf.int32)
-  model.add_feature(degree_slice)
 
-  membership = Input(shape=(None,), dtype=tf.int32)
-  model.add_feature(membership)
+def graph_conv_model(batch_size, tasks):
+  model = TensorGraph(
+      model_dir=model_dir, batch_size=batch_size, use_queue=False)
+  atom_features = Feature(shape=(None, 75), name="ATOMFEATURES")
+  degree_slice = Feature(shape=(None, 2), dtype=tf.int32, name="DEGREE_SLICE")
+  membership = Feature(shape=(None,), dtype=tf.int32, name="MEMBERSHIP")
 
   deg_adjs = []
-  for i in range(model.min_degree, model.max_degree + 1):
-    deg_adj = Input(shape=(None, i + 1), dtype=tf.int32)
-    model.add_feature(deg_adj)
+  for i in range(0, 10 + 1):
+    deg_adj = Feature(shape=(None, i + 1), dtype=tf.int32, name="DEGADJ%i" % i)
     deg_adjs.append(deg_adj)
-
-  gc1 = GraphConvLayer(64, activation_fn=tf.nn.relu)
-  model.add_layer(
-      gc1, parents=[atom_features, degree_slice, membership] + deg_adjs)
-
-  batch_norm1 = BatchNormLayer()
-  model.add_layer(batch_norm1, parents=[gc1])
-
-  gp1 = GraphPoolLayer()
-  model.add_layer(
-      gp1, parents=[batch_norm1, degree_slice, membership] + deg_adjs)
-
-  gc2 = GraphConvLayer(64, activation_fn=tf.nn.relu)
-  model.add_layer(gc2, parents=[gp1, degree_slice, membership] + deg_adjs)
-
-  batch_norm2 = BatchNormLayer()
-  model.add_layer(batch_norm2, parents=[gc2])
-
-  gp2 = GraphPoolLayer()
-  model.add_layer(
-      gp2, parents=[batch_norm2, degree_slice, membership] + deg_adjs)
-
-  dense = Dense(out_channels=128, activation_fn=None)
-  model.add_layer(dense, parents=[gp2])
-
-  batch_norm3 = BatchNormLayer()
-  model.add_layer(batch_norm3, parents=[dense])
-
-  gg1 = GraphGather(batch_size=batch_size, activation_fn=tf.nn.tanh)
-  model.add_layer(
-      gg1, parents=[batch_norm3, degree_slice, membership] + deg_adjs)
+  gc1 = GraphConvLayer(
+      64,
+      activation_fn=tf.nn.relu,
+      in_layers=[atom_features, degree_slice, membership] + deg_adjs,
+      name="CONV1")
+  batch_norm1 = BatchNormLayer(in_layers=[gc1], name="BATCHNORM1")
+  gp1 = GraphPoolLayer(
+      in_layers=[batch_norm1, degree_slice, membership] + deg_adjs,
+      name="POOL1")
+  gc2 = GraphConvLayer(
+      64,
+      activation_fn=tf.nn.relu,
+      in_layers=[gp1, degree_slice, membership] + deg_adjs,
+      name="CONV2")
+  batch_norm2 = BatchNormLayer(in_layers=[gc2], name="BATCHNORM2")
+  gp2 = GraphPoolLayer(
+      in_layers=[batch_norm2, degree_slice, membership] + deg_adjs,
+      name="POOL2")
+  dense = Dense(
+      out_channels=128, activation_fn=None, in_layers=[gp2], name="DENSE1")
+  batch_norm3 = BatchNormLayer(in_layers=[dense], name="BATCHNORM3")
+  gg1 = GraphGather(
+      batch_size=batch_size,
+      activation_fn=tf.nn.tanh,
+      in_layers=[batch_norm3, degree_slice, membership] + deg_adjs,
+      name="GATHER")
 
   costs = []
-  for task in range(num_tasks):
+  labels = []
+  for task in tasks:
```

Comment:
I'm currently more in favor of a utility function in layers which does this loop under the hood.  Its nice knowing that the results are always of the shape (sample, task, value), and you can get a task value which is always an array by looking at model.outputs.